velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp

```diff
@@ -426,47 +436,1466 @@ class MinByAggregate : public MinMaxByAggregate<T, U> {
-std::unique_ptr<exec::Aggregate> create(
+// Similar to NumericMinMaxByAggregate but with a non-numeric value type.
+template <typename U>
+class MinMaxByAggregateWithNonNumericValue : public exec::Aggregate {
+ public:
+  MinMaxByAggregateWithNonNumericValue(TypePtr resultType, U initialValue)
+      : exec::Aggregate(resultType), initialValue_(initialValue) {}
+
+  int32_t accumulatorFixedWidthSize() const override {
+    return sizeof(SingleValueAccumulator) + sizeof(U) + sizeof(bool);
+  }
+
+  void initializeNewGroups(
+      char** groups,
+      folly::Range<const vector_size_t*> indices) override {
+    exec::Aggregate::setAllNulls(groups, indices);
+    for (auto i : indices) {
+      auto group = groups[i];
+      comparisonValue(group) = initialValue_;
+      new (groups[i] + valueAccumulatorOffset()) SingleValueAccumulator();
+      valueIsNull(group) = true;
+    }
+  }
+
+  void finalize(char** /* unused */, int32_t /* unused */) override {}
+
+  void extractValues(char** groups, int32_t numGroups, VectorPtr* result)
+      override {
+    VELOX_CHECK(result);
+    (*result)->resize(numGroups);
+    uint64_t* rawNulls = getRawNulls(result->get());
+
+    for (auto i = 0; i < numGroups; ++i) {
+      char* group = groups[i];
+      if (isNull(group) || valueIsNull(group)) {
+        (*result)->setNull(i, true);
+      } else {
+        clearNull(rawNulls, i);
+        auto* accumulator = valueAccumulator(group);
+        VELOX_DCHECK(accumulator->hasValue());
+        accumulator->read(*result, i);
+      }
+    }
+  }
+
+  void extractAccumulators(char** groups, int32_t numGroups, VectorPtr* result)
+      override {
+    auto rowVector = (*result)->as<RowVector>();
+    auto valueVector = rowVector->childAt(0);
+    auto comparisonVector = rowVector->childAt(1)->asFlatVector<U>();
+
+    rowVector->resize(numGroups);
+    valueVector->resize(numGroups);
+    comparisonVector->resize(numGroups);
+    uint64_t* rawNulls = getRawNulls(rowVector);
+
+    U* rawComparisonValues = comparisonVector->mutableRawValues();
+    BufferPtr nulls = valueVector->mutableNulls(rowVector->size());
+    uint64_t* nullValues = nulls->asMutable<uint64_t>();
+    for (int32_t i = 0; i < numGroups; ++i) {
+      char* group = groups[i];
+      if (isNull(group)) {
+        rowVector->setNull(i, true);
+      } else {
+        clearNull(rawNulls, i);
+        bits::setNull(nullValues, i, valueIsNull(group));
+        if (!valueIsNull(group)) {
+          auto* accumulator = valueAccumulator(group);
+          VELOX_DCHECK(accumulator->hasValue());
+          accumulator->read(valueVector, i);
+        }
+        rawComparisonValues[i] = comparisonValue(group);
+      }
+    }
+  }
+
+  void destroy(folly::Range<char**> groups) override {
+    for (auto group : groups) {
+      valueAccumulator(group)->destroy(allocator_);
+    }
+  }
+
+ protected:
+  template <typename MayUpdate>
+  void addRawInput(
```

Comment:
I wonder if this is a copy-paste from the original class. If so, is there a way to re-use the logic by provide a different updateValues lambda?